### DISCUSSION TITLE
Build is recursively adding more build directories during build

### DISCUSSION THREAD

**Discussion by Rexeh** [marcelotduarte/cx_Freeze#2221] [Q&A]
An odd one, as have had CX_FREEZE in my project for years. Now when I build I'm getting recursive build directories added.

I've stripped back to basics to identify cause but still occurring which makes me think perhaps further config is now required?

Setup.py
```
import sys

from cx_Freeze import Executable, setup

BASE = "Win32GUI" if sys.platform == "win32" else None
TARGET_NAME = "joystick_diagrams.exe"

setup(
    executables=[
        Executable(
            "joystick_diagrams/__main__.py",
            base=BASE,
            target_name=TARGET_NAME,
        )
    ],
)

```

Using TOML and [tool.distutils.build_exe] is now empty to try debug. 

Has anyone seen this behavior before? Below is an output from stdout where you can  see the behaviour begin

copying data from package asttokens...
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\asttokens
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\asttokens\py.typed -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\asttokens\py.typed
copying data from package asyncio...
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\asyncio
copying data from package black...
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\brackets.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\brackets.cp311-win_amd64.pyd        
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\cache.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\cache.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\comments.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\comments.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\const.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\const.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\handle_ipynb_magics.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\handle_ipynb_magics.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\linegen.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\linegen.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\lines.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\lines.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\mode.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\mode.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\nodes.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\nodes.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\numerics.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\numerics.cp311-win_amd64.pyd        
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\parsing.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\parsing.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\py.typed -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\py.typed
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\ranges.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\ranges.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\rusty.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\rusty.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\strings.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\strings.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\trans.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\trans.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\_width_table.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\_width_table.cp311-win_amd64.pyd
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\__init__.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\__init__.cp311-win_amd64.pyd        
copying D:\Git Repos\joystick-diagrams\.venv\Lib\site-packages\black\__init__.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\__init__.cp311-win_amd64.pyd
copying data from package build...
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11
copying D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\frozen_application_license.txt -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\frozen_application_license.txt
copying D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\joystick_diagrams.exe -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\joystick_diagrams.exe
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\astroid
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\astroid\brain
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\astroid\interpreter
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\astroid\interpreter\_import
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\astroid\nodes
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\astroid\nodes\scoped_nodes
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\asttokens
copying D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\asttokens\py.typed -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\asttokens\py.typed      
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\asyncio
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\black
copying D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\black\brackets.cp311-win_amd64.pyd -> D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11\lib\black\brackets.cp311-win_amd64.pyd

**Comment by Rexeh** [comment]
After some debugging, it appears that on first build it works, but from then on the build directory is picked up recursively and added to the build. This seems unexpected as never had this before, has something changed in later versions?

**Comment by marcelotduarte** [comment]
This is really unexpected because the directory should be cleaned:
https://github.com/marcelotduarte/cx_Freeze/blob/main/cx_Freeze/freezer.py#L143-L151
Do you have any changes to PATH or sys.path?
I will try to reproduce.

**Reply by Rexeh** [reply]
Thanks.

My windows PATH contains only the default installations, I'm using poetry and venvs in my project, but not executing this via poetry run only using the venv python installation.

In terms of sys.path overrides/additions no I don't use these.

**Comment by Rexeh** [comment]
@marcelotduarte very weirdly, this is now all working fine, the build directory is cleaned each time as expected and not getting recursion anymore. If you're unable to reproduce might be worth closing, and as I add more config back I'll see if I can trigger it again.

It started when I began moving elements from the setup.py to the TOML, but I am using dynamic versions [tool.setuptools.packages.find] and bunch of other things.

**Reply by marcelotduarte** [reply]
As cx_Freeze depends on setuptools, some settings may negatively affect it, and this should be the case. When they introduced package discovery, it broke down and I think packages find is linked to that. Then calmly, I saw a line in your log that indicates that as you must have the build package installed, when searching for it, it looked for the build directory.
```
copying data from package build...
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build
creating directory D:\Git Repos\joystick-diagrams\build\exe.win-amd64-3.11\lib\build\exe.win-amd64-3.11
```

In this case a simpler solution is the use of `excludes`.

**Comment by marcelotduarte** [comment]
I'll close this, and feel free to contact me again.
